posts/loggingPython.md

```diff
@@ -0,0 +1,291 @@
+
+handler = logging.FileHandler('myLogs.log')
+handler.setLevel(logging.INFO)
+
+logger.addHandler(handler)
+logger.info('You can find this written in myLogs.log')
+```
+
+### Logging to the Cloud
+
+Writing your logs to the cloud seriously makes your life so much easier, it serves as a layer of abstraction so you don’t have to worry how the logs are getting to the service and you can focus your time on what’s important.
+
+_Disclaimer: I’m a current employee @ Timber. This section is entirely optional, but I seriously believe that logging to the cloud will make your life easier (and you can try it for completely free)._
+
+```bash
+pip install timber
+```
+
+```python
+import logging
+import timber
+
+logger = logging.getLogger(__name__)
+
+timber_handler = timber.TimberHandler(api_key='...')
+logger.addHandler(timber_handler)
+```
+
+**That’s it.** All you need to do is get your API key from [timber.io](https://timber.io/) and you’ll be able to see your logs. We automatically capture them from the logging module, so you can continue to follow best-practices and log normally, while we seamlessly add context. 
+
+### Creating the _Picasso_ of Logs
+ 
+Like a great painting, your logs need to paint a picture for the next person who looks at them. Most people dump the object or a breakpoint in their logs, but they’ll hate themselves in 2 years when they try to understand what was happening in their code. The logs _must_ give the next developer a birds-eye-view into the actions taken by the user up to a point in time. 
+
+A generic log message provides just about as little information as no log message. Imagine if you had to go through your logs and you saw `purchase completed`. This doesn’t help you answer any questions. *When was the purchase completed? Who completed it? What did they buy?*
+
+This can all be done by formatting your logs:
+```python
+logFormatter = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'
+logging.basicConfig(format=logFormatter)
+```
+
+#### Structured Data
+
+The _biggest_ issue we’ve seen is the pollution of log messages. People will put objects or variables into the log. Though it’s helpful while debugging your code, it’s impossible to decipher what you were thinking when looking at the stream in the future. The best way to deal with this is to push all your metadata to the `extra` object. Using this, you’ll be able to decipher your messages from the stream, while still grab the full context to dig deeper into your user from any messages.
+
+Some of this data, such as `time` and `levelname` can be captured automatically, but if you can (and should) push `extra` context to your logs.
+
+```python
+logger = logging.getLogger(__name__)
+logger.info('purchase completed', extra={‘user': 'Sid Panjwani'})
+```
+
+If you’re wondering why you do not see your logs, remember to change the `level`. By default, it’s set to `warning`.
+
+*Yep, it’s that’s easy…*
+
+### Tracking Exceptions through Logs
```

Comment:
This wasn't in the template, but I think it's still a pretty interesting section.

Reply:
I'd recommend getting rid of this section entirely. The way it's presented does not provide an overarching exception redirection methodology.